sql code/ksm-matching-gifts.sql

```diff
@@ -0,0 +1,216 @@
+                  HH.yesterday,
+                  HH.curr_fy
+,ROW_NUMBER() OVER(PARTITION BY hh.HOUSEHOLD_ID ORDER BY hh.DATE_OF_RECORD DESC)RW
+FROM rpt_pbh634.v_ksm_giving_trans_hh HH
+cross join rpt_pbh634.v_current_calendar cal
+ where (hh.fiscal_year = cal.CURR_FY - 0 
+or hh.fiscal_year = cal.CURR_FY - 1)
+and hh.TX_GYPM_IND = 'M'
+ ),
+
+ 
+--- Last 4 Matching Gifts
+ 
+GI AS( select house.HOUSEHOLD_ID
+     ,max(decode(RW,1,house.TRANSACTION_TYPE)) TRANSACTION_TYPE1
+     ,max(decode(RW,1,house.DATE_OF_RECORD)) DATE1
+     ,max(decode(RW,1,house.TX_NUMBER)) TX_NUMBER1
+     ,max(decode(RW,1,house.HH_CREDIT)) CREDIT1
+     ,max(decode(RW,1,house.ALLOC_SHORT_NAME)) ALLOCATION1
+     ,max(decode(RW,2,house.TRANSACTION_TYPE)) TRANSACTION_TYPE2
+     ,max(decode(RW,2,house.DATE_OF_RECORD)) DATE2
+     ,max(decode(RW,2,house.TX_NUMBER)) TX_NUMBER2
+     ,max(decode(RW,2,house.ALLOC_SHORT_NAME)) ALLOCATION2
+     ,max(decode(RW,2,house.HH_CREDIT)) CREDIT2
+     ,max(decode(RW,3,house.TRANSACTION_TYPE)) TRANSACTION_TYPE3
+     ,max(decode(RW,3,house.DATE_OF_RECORD)) DATE3
+     ,max(decode(RW,3,house.TX_NUMBER)) TX_NUMBER3
+     ,max(decode(RW,3,house.ALLOC_SHORT_NAME)) ALLOCATION3
+     ,max(decode(RW,3,house.HH_CREDIT)) CREDIT3
+     ,max(decode(RW,4,house.TRANSACTION_TYPE)) TRANSACTION_TYPE4
+     ,max(decode(RW,4,house.DATE_OF_RECORD)) DATE4
+     ,max(decode(RW,4,house.TX_NUMBER)) TX_NUMBER4
+     ,max(decode(RW,4,house.ALLOC_SHORT_NAME)) ALLOCATION4
+     ,max(decode(RW,4,house.HH_CREDIT)) CREDIT4
+ from GIVING_TRANS house
+ group by HOUSE.HOUSEHOLD_ID)
+
+
+Select h.ID_NUMBER,
+h.HOUSEHOLD_ID,
+h.REPORT_NAME,
+h.FIRST_KSM_YEAR,
+h.PROGRAM,
+h.PROGRAM_GROUP,
+h.SPOUSE_ID_NUMBER,
+h.SPOUSE_REPORT_NAME,
+h.SPOUSE_SUFFIX,
+h.SPOUSE_FIRST_KSM_YEAR,
+h.SPOUSE_PROGRAM,
+h.SPOUSE_PROGRAM_GROUP,
+h.HOUSEHOLD_PRIMARY,
+a.prospect_manager,
+a.lgos,
+s.NO_CONTACT,
+s.ACTIVE_WITH_RESTRICTIONS,
+s.NO_EMAIL_IND,
+s.NO_EMAIL_SOL_IND,
+spouse.NO_CONTACT spouse_no_contact_flag,
+spouse.ACTIVE_WITH_RESTRICTIONS spouse_active_withrestrictions,
+spouse.NO_EMAIL_IND as spouse_no_email_ind,
+spouse.NO_EMAIL_SOL_IND as spouse_no_email_sol_ind,
+p.Address_Type,
+p.care_of,
+p.Street1,
+p.Street2,
+p.Street3,
+p.City,
+p.State_code,
+p.Zipcode,
+p.Country,
+GI.TRANSACTION_TYPE1,
+GI.DATE1,
+GI.TX_NUMBER1,
+GI.CREDIT1,
+GI.ALLOCATION1,
+GI.TRANSACTION_TYPE2,
+GI.DATE2,
+GI.TX_NUMBER2,
+GI.CREDIT2,
+GI.ALLOCATION2,
+GI.TRANSACTION_TYPE3,
+GI.DATE3,
+GI.TX_NUMBER3,
+GI.CREDIT3,
+GI.ALLOCATION3,
+GI.TRANSACTION_TYPE4,
+GI.DATE4,
+GI.TX_NUMBER4,
+GI.CREDIT4,
+GI.ALLOCATION4
+from h
+inner join d 
+     on d.id_number = h.id_number
+left join  a 
+     on a.id_number = h.id_number  
+left join s 
+     on s.id_number = h.id_number
+--- special handling code, but for spouses
+left join  spouse 
+     on spouse.SPOUSE_ID_NUMBER = h.SPOUSE_ID_NUMBER
+left join p
+     on p.id_number = h.id_number
+inner join GI
+     ON H.ID_NUMBER = GI.HOUSEHOLD_ID
```

Comment:
ON H.ID_NUMBER = GI.ID_NUMBER
-- not household id
-- Then add a row_number() (partition by household_id) so that you can filter for row_number = 1 in Tableau, to leave out spouses

Reply:
Changes:

- Inner Join on ID NUMBER (H.ID_number = GI.ID_NUMBER)
- Partitioned in the final query by household ID and ordered by household ID (Named it as Spouse Exclusion)